File: cascadingFailuresTaskMigration/src/main/java/opt/OptMigration.java
package opt;

import evaluation.Evalution;
import input.Group;
import input.MigrationRecord;
import input.Robot;
import input.Task;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class OptMigration {
    ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath;
    HashMap<Integer, Group> idToGroups;
    HashMap<Integer, Robot> idToRobots;
    Double a;
    Double b;
    HashMap<Integer,Robot> bestIdRobots ;
    HashMap<Integer,Group> bestIdGroups ;
    Double minTargetValue = Double.MAX_VALUE;
    HashMap<Task,Robot> taskToRobot ;
    List<Task> allTasks = new ArrayList<>();
    public OptMigration(ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath,
                        HashMap<Integer, Group> idToGroups, HashMap<Integer, Robot> idToRobots,
                        Double a, Double b) {
        this.shortestPath = shortestPath;
        this.idToGroups = idToGroups;
        this.idToRobots = idToRobots;
        this.a = a;
        this.b = b;
    }

    public List<MigrationRecord> run() {
        for (Integer robotId : idToRobots.keySet()) {
            Robot robot = idToRobots.get(robotId);
            allTasks.addAll(robot.getTasksList());
        }

        HashMap<Integer,Robot> idToRobotsTemp =  robotMapCopyOrg(idToRobots);
        HashMap<Integer, Group> idToGroupsTemp = groupMapCopyOrg(idToGroups);
        taskToRobot = new HashMap<>();

        for (Integer robotId : idToRobots.keySet()) {
            Robot robot = idToRobots.get(robotId);
            List<Task> tasksList = robot.getTasksList();
            for (Task task : tasksList) {
                taskToRobot.put(task,robot);
            }
        }

        bestIdRobots = robotMapCopy(idToRobots);
        bestIdGroups = groupMapCopy(idToGroups);

        backtrace(idToRobotsTemp,idToGroupsTemp,allTasks,0);

        //用计算迁移的函数来返回
        List<MigrationRecord> ret;
        ret = calculateMigrationRecord(bestIdRobots);
        return ret;
    }



    private void backtrace(HashMap<Integer, Robot> idToRobotsTemp, HashMap<Integer, Group> idToGroupsTemp,List<Task> allTasks,int index) {
        if (index==allTasks.size()) {
            Double targetValue = calculateTargetValue(idToRobotsTemp,idToGroupsTemp);
            if (targetValue<minTargetValue) {
                bestIdGroups = groupMapCopy(idToGroupsTemp);
                bestIdRobots = robotMapCopy(idToRobotsTemp);
                minTargetValue = targetValue;
                System.out.println(index+" "+minTargetValue);
            }
            return;
        }
        for (Integer id : idToRobotsTemp.keySet()) {
            Robot robotTemp = idToRobotsTemp.get(id);
            if (robotTemp.getFaultA()==1) {
                continue;
                //有故障的节点不能够接收任务
            }
            for (int i = index;i<allTasks.size();i++) {
                Task task = allTasks.get(i);
                List<Task> tasksList = robotTemp.getTasksList();
                tasksList.add(task);
                robotTemp.setTasksList(tasksList);
                robotTemp.setLoad(robotTemp.getLoad()+task.getSize());
                backtrace(idToRobotsTemp,idToGroupsTemp,allTasks,index+1);
                tasksList.remove(task);
                robotTemp.setLoad(robotTemp.getLoad()-task.getSize());
            }
        }
    }

    private Double calculateTargetValue(HashMap<Integer, Robot> idToRobotsTemp, HashMap<Integer, Group> idToGroupsTemp) {
        List<MigrationRecord> tempMigrationRecords =calculateMigrationRecord(idToRobotsTemp);
        Evalution evalution = new Evalution(idToRobotsTemp,idToGroupsTemp);
        Double survivalRate = evalution.calculateMeanSurvivalRate(new ArrayList<>(idToRobotsTemp.values()));
        Double executeTasksCost = evalution.calculateExecuteTasksCost(new ArrayList<>(idToRobotsTemp.values()));
        Double migrationCost = evalution.calculateMigrationCost(shortestPath, tempMigrationRecords);
        return a*(executeTasksCost+migrationCost)-b*survivalRate;
    }

    private List<MigrationRecord> calculateMigrationRecord(HashMap<Integer, Robot> idToRobotsTemp) {
        List<MigrationRecord> records = new ArrayList<>();
        for (Integer id : idToRobots.keySet()) {
            Robot robotTemp = idToRobotsTemp.get(id);
            for (Task task : robotTemp.getTasksList()) {
                if (taskToRobot.keySet().contains(task)) {
                    Robot robot = taskToRobot.get(task);
                    if(robot.getRobotId()!=id) {
                        MigrationRecord record = new MigrationRecord();
                        record.setFrom(robot.getRobotId());
                        record.setTo(id);
                        records.add(record);
                    }
                }
            }
        }
        return records;
    }


    private HashMap<Integer, Robot> robotMapCopyOrg(HashMap<Integer, Robot> idToRobots) {
        HashMap<Integer,Robot> idToRobotsTemp =  new HashMap<>();
        for (Integer id : idToRobots.keySet()) {
            Robot robot = idToRobots.get(id);
            Robot robotTemp = new Robot();
            robotTemp.setLoad(0);
            robotTemp.setTasksList(new ArrayList<>());
            robotTemp.setGroupId(robot.getGroupId());
            robotTemp.setCapacity(robot.getCapacity());
            robotTemp.setRobotId(id);
            idToRobotsTemp.put(id,robotTemp);
        }
        return idToRobotsTemp;
    }

    private HashMap<Integer, Group> groupMapCopyOrg(HashMap<Integer, Group> idToGroups) {
        HashMap<Integer,Group> idToGroupsTemp = new HashMap<>();
        for (Integer groupId : idToGroups.keySet()) {
            Group group = idToGroups.get(groupId);
            Group groupTemp = new Group();
            groupTemp.setGroupLoad(0);
            groupTemp.setGroupCapacity(group.getGroupCapacity());
            groupTemp.setLeader(group.getLeader());
            groupTemp.setGroupId(groupId);
            groupTemp.setRobotIdInGroup(group.getRobotIdInGroup());
            idToGroupsTemp.put(groupId,groupTemp);
        }
        return idToGroupsTemp;
    }


    private HashMap<Integer, Robot> robotMapCopy(HashMap<Integer, Robot> idToRobots) {
        HashMap<Integer,Robot> idToRobotsTemp =  new HashMap<>();
        for (Integer id : idToRobots.keySet()) {
            Robot robot = idToRobots.get(id);
            Robot robotTemp = new Robot();
            robotTemp.setLoad(robot.getLoad());
            robotTemp.setTasksList(robot.getTasksList());
            robotTemp.setGroupId(robot.getGroupId());
            robotTemp.setCapacity(robot.getCapacity());
            robotTemp.setRobotId(id);
            idToRobotsTemp.put(id,robotTemp);
        }
        return idToRobotsTemp;
    }

    private HashMap<Integer, Group> groupMapCopy(HashMap<Integer, Group> idToGroups) {
        HashMap<Integer,Group> idToGroupsTemp = new HashMap<>();
        for (Integer groupId : idToGroups.keySet()) {
            Group group = idToGroups.get(groupId);

            Group groupTemp = new Group();
            groupTemp.setGroupLoad(group.getGroupLoad());
            groupTemp.setGroupCapacity(group.getGroupCapacity());
            groupTemp.setLeader(group.getLeader());
            groupTemp.setGroupId(groupId);
            groupTemp.setRobotIdInGroup(group.getRobotIdInGroup());
            idToGroupsTemp.put(groupId,groupTemp);
        }
        return idToGroupsTemp;
    }
}

File: cascadingFailuresTaskMigration_python/opt/opt_migration.py
import copy
from typing import Dict, List
from input.group import Group
from input.robot import Robot
from input.task import Task
from input.migration_record import MigrationRecord
from evaluation.evalution import Evalution


class OptMigration:
    def __init__(self, shortest_path_dict: Dict, id_to_groups: Dict[int, Group],
                 id_to_robots: Dict[int, Robot], a: float, b: float):
        self.shortest_path_dict = shortest_path_dict
        self.id_to_groups = id_to_groups
        self.id_to_robots = id_to_robots
        self.a = a
        self.b = b
        self.best_id_robots = None
        self.best_id_groups = None
        self.min_target_value = float('inf')
        self.task_to_robot = {}
        self.all_tasks = []

    def run(self) -> List[MigrationRecord]:
        """Run optimization algorithm to find best task allocation."""
        # Collect all tasks
        for robot_id in self.id_to_robots.keys():
            robot = self.id_to_robots[robot_id]
            self.all_tasks.extend(robot.tasks_list)

        # Create temporary copies
        id_to_robots_temp = self._robot_map_copy_org(self.id_to_robots)
        id_to_groups_temp = self._group_map_copy_org(self.id_to_groups)

        # Record original task-robot mapping
        for robot_id in self.id_to_robots.keys():
            robot = self.id_to_robots[robot_id]
            for task in robot.tasks_list:
                self.task_to_robot[task] = robot

        self.best_id_robots = self._robot_map_copy(self.id_to_robots)
        self.best_id_groups = self._group_map_copy(self.id_to_groups)

        self._backtrace(id_to_robots_temp, id_to_groups_temp, self.all_tasks, 0)

        # Calculate migration records based on best allocation
        ret = self._calculate_migration_record(self.best_id_robots)
        return ret

    def _backtrace(self, id_to_robots_temp: Dict, id_to_groups_temp: Dict,
                  all_tasks: List[Task], index: int):
        """Backtracking algorithm to find optimal task allocation."""
        if index == len(all_tasks):
            target_value = self._calculate_target_value(id_to_robots_temp, id_to_groups_temp)
            if target_value < self.min_target_value:
                self.best_id_groups = self._group_map_copy(id_to_groups_temp)
                self.best_id_robots = self._robot_map_copy(id_to_robots_temp)
                self.min_target_value = target_value
                print(f"{index} {self.min_target_value}")
            return

        for robot_id in id_to_robots_temp.keys():
            robot_temp = id_to_robots_temp[robot_id]
            if robot_temp.fault_a == 1:
                continue  # Faulty robots cannot receive tasks

            for i in range(index, len(all_tasks)):
                task = all_tasks[i]
                tasks_list = robot_temp.tasks_list
                tasks_list.append(task)
                robot_temp.tasks_list = tasks_list
                robot_temp.load = robot_temp.load + task.size

                self._backtrace(id_to_robots_temp, id_to_groups_temp, all_tasks, index + 1)

                tasks_list.remove(task)
                robot_temp.load = robot_temp.load - task.size

    def _calculate_target_value(self, id_to_robots_temp: Dict, id_to_groups_temp: Dict) -> float:
        """Calculate target optimization value."""
        temp_migration_records = self._calculate_migration_record(id_to_robots_temp)
        evalution = Evalution(id_to_robots_temp, id_to_groups_temp)

        survival_rate = evalution.calculate_mean_survival_rate(list(id_to_robots_temp.values()))
        execute_tasks_cost = evalution.calculate_execute_tasks_cost(list(id_to_robots_temp.values()))
        migration_cost = evalution.calculate_migration_cost(self.shortest_path_dict, temp_migration_records)

        return self.a * (execute_tasks_cost + migration_cost) - self.b * survival_rate

    def _calculate_migration_record(self, id_to_robots_temp: Dict) -> List[MigrationRecord]:
        """Calculate migration records by comparing with original allocation."""
        records = []
        for robot_id in self.id_to_robots.keys():
            robot_temp = id_to_robots_temp[robot_id]
            for task in robot_temp.tasks_list:
                if task in self.task_to_robot:
                    robot = self.task_to_robot[task]
                    if robot.robot_id != robot_id:
                        record = MigrationRecord()
                        record.from_robot = robot.robot_id
                        record.to_robot = robot_id
                        records.append(record)
        return records

    def _robot_map_copy_org(self, id_to_robots: Dict) -> Dict:
        """Create copy of robot map with reset loads."""
        id_to_robots_temp = {}
        for robot_id, robot in id_to_robots.items():
            robot_temp = Robot()
            robot_temp.load = 0.0
            robot_temp.tasks_list = []
            robot_temp.group_id = robot.group_id
            robot_temp.capacity = robot.capacity
            robot_temp.robot_id = robot_id
            robot_temp.fault_a = robot.fault_a
            robot_temp.fault_o = robot.fault_o
            robot_temp.fault_s = robot.fault_s
            id_to_robots_temp[robot_id] = robot_temp
        return id_to_robots_temp

    def _group_map_copy_org(self, id_to_groups: Dict) -> Dict:
        """Create copy of group map with reset loads."""
        id_to_groups_temp = {}
        for group_id, group in id_to_groups.items():
            group_temp = Group()
            group_temp.group_load = 0.0
            group_temp.group_capacity = group.group_capacity
            group_temp.leader = group.leader
            group_temp.group_id = group_id
            group_temp.robot_id_in_group = group.robot_id_in_group.copy()
            id_to_groups_temp[group_id] = group_temp
        return id_to_groups_temp

    def _robot_map_copy(self, id_to_robots: Dict) -> Dict:
        """Create full copy of robot map."""
        id_to_robots_temp = {}
        for robot_id, robot in id_to_robots.items():
            robot_temp = Robot()
            robot_temp.load = robot.load
            robot_temp.tasks_list = robot.tasks_list.copy()
            robot_temp.group_id = robot.group_id
            robot_temp.capacity = robot.capacity
            robot_temp.robot_id = robot_id
            robot_temp.fault_a = robot.fault_a
            robot_temp.fault_o = robot.fault_o
            robot_temp.fault_s = robot.fault_s
            id_to_robots_temp[robot_id] = robot_temp
        return id_to_robots_temp

    def _group_map_copy(self, id_to_groups: Dict) -> Dict:
        """Create full copy of group map."""
        id_to_groups_temp = {}
        for group_id, group in id_to_groups.items():
            group_temp = Group()
            group_temp.group_load = group.group_load
            group_temp.group_capacity = group.group_capacity
            group_temp.leader = group.leader
            group_temp.group_id = group_id
            group_temp.robot_id_in_group = group.robot_id_in_group.copy()
            id_to_groups_temp[group_id] = group_temp
        return id_to_groups_temp
